student/src/main/java/tds/student/web/filter/ConcurrentRequestLimitFilter.java

```diff
@@ -0,0 +1,53 @@
+package tds.student.web.filter;
+
+import org.apache.http.HttpHeaders;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.stereotype.Component;
+import org.springframework.web.filter.OncePerRequestFilter;
+
+import javax.servlet.FilterChain;
+import javax.servlet.ServletException;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import java.io.IOException;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import static java.lang.Integer.parseInt;
+
+/**
+ * This concurrent request limiting filter will fast-fail with a 503 if the maximum
+ * number of concurrent requests are already being resolved.
+ */
+@Component
+public class ConcurrentRequestLimitFilter extends OncePerRequestFilter {
```

Comment:
This filter will fast-fail a request with a 503 and a "Retry-After" header if we're already serving a configurable number of concurrent requests.